contracts/Proxy.sol

```diff
@@ -17,19 +17,23 @@ contract Proxy is Proxied {
     }
 
     /// @dev Fallback function forwards all transactions and returns all received return data.
-    function() external payable {
+    fallback() external payable {
         address _masterCopy = masterCopy;
         assembly {
-            calldatacopy(0, 0, calldatasize)
-            let success := delegatecall(not(0), _masterCopy, 0, calldatasize, 0, 0)
-            returndatacopy(0, 0, returndatasize)
+            calldatacopy(0, 0, calldatasize())
+            let success := delegatecall(not(0), _masterCopy, 0, calldatasize(), 0, 0)
+            returndatacopy(0, 0, returndatasize())
             switch success
                 case 0 {
-                    revert(0, returndatasize)
+                    revert(0, returndatasize())
                 }
                 default {
-                    return(0, returndatasize)
+                    return(0, returndatasize())
                 }
         }
     }
+
+    receive() external payable {
+        revert();
+    }
```

Comment:
Do we really want this behaviour? It seem like e.g. the Safe (using the Proxy contract) should be able to receive ether without reverting.

Reply:
What should we do otherwise? Just leave it as an empty method?

Reply:
I think so, or else I imagine sending Eth to your Safe (if it uses this proxy class) would revert.

Reply:
I think this implementation is incorrect. If this `Proxy` implementation would be used by the `Safe` for example, it would revert when someone sent you Eth, which I imagine is not what you want. I suspect the correct behaviour is to proxy the call to the `_masterCopy` as well, just without calldata.

The easier approach, would be to just remove the `receive` function as according to the docs <https://docs.soliditylang.org/en/v0.6.12/contracts.html#fallback-function> (emphasis my own):
> The `fallback` function [..] is executed on a call to the contract if none of the other functions match the given function signature, or if no data was supplied at all **and there is no receive Ether function**.

Reply:
Leaving it empty also does not work, as some contracts (e.g. WETH) execute logic when they are sent ETH without any additional calldata. If the method was empty it would not trigger this logic on the proxied contract (as no delegatecall would be executed). I was under the assumption that just omitting `receive` would not work because of some solidity warning but agree that this would be the best option if possible. Otherwise I guess you'd have to copy some of the code from the fallback method to also invoke receive on the proxied contract.

Reply:
According to the docs, omitting the the `receive` function entirely should work. Maybe there is some way to silence the warnings if they show up?

Reply:
There was a compiler warning that I should include it. This is why I have included it.

```
gnosis/util-contracts/contracts/Proxy.sol:11:1: Warning: This contract has a payable fallback function, but no receive ether function. Consider adding a receive ether function.
contract Proxy is Proxied {
^ (Relevant source part starts here and spans across multiple lines).
gnosis/util-contracts/contracts/Proxy.sol:20:5: The payable fallback function is defined here.
    fallback() external payable {
    ^ (Relevant source part starts here and spans across multiple lines).
```

Reply:
I can omit it since it is just a warning, but its ugly. However, it does compile just fine.

Reply:
@nlordell

Reply:
Maybe we can take @fleupold's suggestion:
> Otherwise I guess you'd have to copy some of the code from the fallback method to also invoke receive on the proxied contract.

Reply:
Yea, I could do that but I'm not entirely sure what "some" of the code to copy over... could we just do something like this?

```js
    receive() external payable {
        fallback();
    }
```

Reply:
I am also totally fine with the warning for now however annoying it may be.

Reply:
> could we just do something like this?

If the compiler lets you, I think that is my favorite approach.

Reply:
Unfortunately this wont work, but I have discovered, we can use receive as fallback and simply not include fallback without compiler warning. I think this is the best way to go.

Reply:
If I'm not mistaken, that might cause calls with calldata to simply revert: 
> It is executed on a call to the contract with empty calldata

Basically, it won't proxy any contract calls.

https://docs.soliditylang.org/en/v0.6.12/contracts.html?highlight=receive#receive-ether-function

Reply:
Ok, in that case I think we will have to live with the warning. I was not able to call fallback from within receive...

Reply:
When in doubt, we could always do what OZ does in this regard: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Proxy.sol


I'm fine with either not implementing it our using it the way OZ does (with a slight preference for not having to see warnings on every compilation). However I think it's important to understand that this call needs to be **forwarded** to the underlying contract so that that contract can implement logic when receiving ether (e.g. [WETH](https://github.com/gnosis/canonical-weth/blob/master/contracts/WETH9.sol#L32) will call `deposit()` in that case). Leaving it empty or reverting will break the proxy pattern.

Reply:
This is a great suggestion. Its unfortunate that such a work around is necessary, but its fairly clean. I also prefer not to see warnings! So I will do something like this 

```
    function _fallback() internal {
        address _masterCopy = masterCopy;
        assembly {
            calldatacopy(0, 0, calldatasize())
            let success := delegatecall(not(0), _masterCopy, 0, calldatasize(), 0, 0)
            returndatacopy(0, 0, returndatasize())
            switch success
                case 0 {
                    revert(0, returndatasize())
                }
                default {
                    return(0, returndatasize())
                }
        }
    }

    fallback () external payable {
        _fallback();
    }

    receive () external payable {
        _fallback();
    }
```

in both this PR and solc-7.


Does this mean that we are not happy using just receive without fallback? This also compiles without warnings, but perhaps I have misunderstood the details. Clearly receive and fallback can not simply be used interchangeably since there is now a dedicated distinction between them. But what I have included above is simply making both receive and fallback do exactly the same thing (as far as I understand).

Reply:
After looking closer at the OZ contract, it looks like perhaps we should simply be using their Proxy contracts instead of maintaining this one. Otherwise, it seems we are just doing the same thing twice.

Reply:
> Does this mean that we are not happy using just receive without fallback?

As Nick suggested this is likely going to cause problems if you are trying to proxy a regular method (one that has calldata like ERC20.transfer(to, amount)). If we did this, I'd like to see a unit test making sure this still works.

> it looks like perhaps we should simply be using their Proxy contracts instead of maintaining this one

I think this is a good suggestion. I'm not sure if there are other than historic reasons that we built our own. It is however not really related to this PR and should be separate (could be done before if you want to avoid upgrading this piece).

Reply:
The proxy is from the Safe project and we wrote our own because the one from OZ uses up more gas

Reply:
Right, this is a good idea. However, since the purpose of this PR is merely to upgrade the contracts, I will create an issue to remove the Proxy separately.

Reply:
https://github.com/gnosis/util-contracts/issues/43